\documentclass[11pt, a4paper,oneside,chapterprefix=false]{scrbook}

\usepackage{a4wide}
\usepackage{times}
\usepackage{helvet}   % sets sans serif font

\usepackage{amsmath,amssymb,amsthm}

\usepackage{graphicx}
\usepackage{subfigure}  
\usepackage{fancybox} % for shadowed or double bordered boxes
\usepackage{fancyhdr}
\usepackage[dvipsnames]{xcolor}
\usepackage{rotating}

\usepackage{hyperref}
\hypersetup{
    colorlinks=true,
    linkcolor=blue,
    filecolor=magenta,      
    urlcolor=cyan,
}

\DeclareGraphicsExtensions{.pdf, .jpg, .png}
\DeclareOldFontCommand{\sf}{\normalfont\sffamily}{\mathsf}

%% macros
\input{include/math}
\input{include/codelisting_layout}

\usepackage{color}
\definecolor{RED}{rgb}{1,0,0}
\definecolor{GREEN}{rgb}{0,0.7,0}
\definecolor{BLUE}{rgb}{0,0,1}
\newcommand{\FIXME}[1]{{\color{RED}{\textbf{FIX}: #1}}}

\addtolength{\textheight}{2.0cm}
\addtolength{\voffset}{-1cm}
\addtolength{\textwidth}{1.8cm}
\addtolength{\hoffset}{-.9cm}

\widowpenalty=10000
\clubpenalty=10000

\newcommand{\todo}[1]{}
\renewcommand{\todo}[1]{{\color{red} TODO: {#1}}}


\lstdefinelanguage{JavaScript2}{
  keywords={typeof, new, true, false, catch, function, return, null, catch, switch, var, if, in, while, do, else, case, break, for, forEach},
  keywordstyle=\color{orange}\bfseries,
  ndkeywords={class, export, boolean, throw, implements, import, this},
  ndkeywordstyle=\color{darkgray}\bfseries,
  identifierstyle=\color{black},
  sensitive=false,
  comment=[l]{//},
  morecomment=[s]{/*}{*/},
  commentstyle=\color{purple}\ttfamily,
  stringstyle=\color{ForestGreen}\ttfamily,
  morestring=[b]',
  morestring=[b]"
}

\begin{document}

\frontmatter
%\maketitle %automatic version
% --- selfmade version ----
\begin{titlepage}
	\setlength{\parindent}{0cm}
	\addtolength{\textheight}{1.0cm}

	\vspace{0.5cm}
	\Huge
	{\textbf{A visualization of biodiversity through time\\ }}

	\vfill \vfill \vfill
	\begin{center}
	\includegraphics*[width=0.8\textwidth]{figures/cover/cover}
    \end{center}
	\vfill
	\sf \Large
	Master Thesis \\
	\today \\[0.5cm]
	\large
	by Wuxing Dong, 18-946-038

	\vfill \vfill \vfill
	\begin{minipage}[b]{0.5\textwidth}
	Supervisors: \\
	Prof. Dr. Renato Pajarola \\
	Gaudenz Halter \\
	\end{minipage}
	%
	\begin{minipage}[b]{0.5\textwidth} \raggedleft
	Visualization and MultiMedia Lab \\
	Department of Informatics \\
	University of Z{\"u}rich
	\end{minipage}

	\vfill
	\hrule
	\vspace{0.5cm}
	\includegraphics*[width=0.3\textwidth]{figures/cover/uzh_logo} \hfill
	\includegraphics*[width=0.3\textwidth]{figures/cover/vmml_logo}
\end{titlepage}
%%

%=====================================================================
\include{abstract}
%=====================================================================

\tableofcontents

\mainmatter


%=====================================================================
\chapter{Introduction} \label{chp:introduction}
%=====================================================================

\section{Effective visualization of data} \label{sec:one}
In the time of the information age, the Internet has become a primary source to educate ourselves about the world. It is therefore a great opportunity for the scientific community to take advantage of the web and create high quality products to communicate their findings and ideas to the non-experts and general public. Textbook-styled verbal explanations can be rigorous, however, they easily become tedious and are already made available by school and college education. Videos are based on visual learning, quick to absorb, yet there is no engagement with the viewers. An interactive data visualization can provide an excellent way for users to understand scientific concepts. New knowledge is conveyed by direct visual impact that easily draw attention, and insights can be formed by actively exploring the data and observing the response of the presentation due to the users' own actions. This learn-by-doing strategy can consolidate the new knowledge. \\

Although graphics plays a major role to deliver scientific messages, there is a lack of formal theory on producing good graphics, and domain experts tend to rely on practical experiences that are guided by certain design principles. Increasing amount of studies have focused on concluding such principles. Chen and colleagues have summarized that three main aspects to pay attention to in a visualization project are its contents, the relevance in the context of the larger whole as well as the forms and aesthetics (construction). Similarly, McCandless argued that a good visualization requires integrity of information, interestingness of story (concept), usefulness of goal and beauty of the visual form.\\

Before embarking on the production of a visualization, one should note the characteristics of human visual perception to improve the design strategies in practice. Cognitive psychological studies aimed at examining human visual attention by exposing users to certain visual processing tasks. The results of those studies suggested some guidelines for more effective visualization. When the visual features of a certain type of data is expected or previously known, the accuracy and speed of visual perception is drastically improved. Therefore, a visual feature should be assigned to the same data dimension, or type of data. It is also important to note that attention is limited and selective, therefore the amount and diversity of information should be kept within the limits of cognitive load. When many categories of data are present, compromises are needed to retain graphics complexity and only demonstrate a few number of categories at a time. \\

Through early years of research on how images are analyzed visually, it has been found that detailed vision is only achievable by actively focusing on small portion of the visual field. However, a few visual features, such as colors and curvatures, are subjected to preattentive processing, which can be effortlessly and rapidly detected by low-level visual processes. Colors is thus a powerful method in graphics. It can encode extra layers of information, and serves a leading factor to make a visualization memorable. Three major schemes for using colors are sequential, diverging and qualitative. A sequential scheme relates increasing value of the data to the intensity of the color, while using only one or two hues. In diverging scheme, two extremes are shown by two sequential schemes, and the middle or neutral value corresponds to a white or neutral color. The qualitative scheme uses a variety of distinguishable colors to represent data belonging to distinct categories. It is a good practice to choose color schemes that are distinguishable enough, such that minimal information is lost when converted into a gray scale. In addition, when a range of colors are used together, reducing the opacity of certain or all colors avoids overwhelming visual impact to the viewers. \\

Design principles have also been examined and summarized by different studies. The design process should be guided by the core messages to deliver. It is important to first prioritize what information to show, envision the final result, and then choose appropriate geometries that fulfill the objectives. Geometries are the forms in which data is illustrated, and the majority of them can be categorized into amounts, compositions, distributions or relationships. Often, different geometries can be considered on the same data, however, the most efficient geometry should be of high data-ink ratio, which is the ratio between the amount of ink allocated on the actual data to the total amount of ink used in a figure. While a high data-ink ratio indicates high efficiency, it is also important for a visualization to be simple, self-explanatory  and straight to the points, such that the viewers are not overwhelmed with too much information. \\


\section{Background on biodiversity and tectonic change} \label{sec: two}
The data this thesis chooses to visualize is biodiversity through earth history. Considering various pressing environmental issues nowadays, it is important to raise the awareness of the general public on how different forms of life came to exist and change during the course of earth history. Climate change has become a non-negligible factor influencing on our daily life and the future of human society. Recently, the world sees increasing amount of extreme weather conditions, causing significant losses in human lives and properties. The ecosystem is also under threat of a mass extinction event, after which 90\% of all species are expected to disappear, and its leading cause is considered to be human activity. 
By understanding the history of life through simple and interesting visualizations, viewers could appreciate the inseparable relationship with our environment, and be more promoted to act in a more environmentally-friendly and sustainable manner.

\subsection{Biodiversity as a result of evolution} \label{subsec: one}
\begin{figure}[h]
	\centering
	\subfigure[John Gould's sketches of the various beak shapes of the Galapagos finches as a result of evolution. ]{\includegraphics[width=0.45\textwidth]{figures/introduction/finches}\label{fig:finches}} 
	\hfill
	\subfigure[Biodiversity during Phanerozoic experiences non linear change with mass extinction events. ]{\includegraphics[width=0.45\textwidth]{figures/introduction/phanerozoic_biodiversity}\label{fig:phanerozoic_biodiversity}} 
	\label{fig:evolution}
\end{figure}
Current views on the process of species formation is based on modern evolution theory originated from Darwin's work. To summarize briefly, all forms of life are argued to share a common ancestor, which diversified over the course of evolution to form the current tree of life. An organism's genomic information is encoded in DNA or RNA, which is replicated over generations. However, recombination and random mutation events occur during the replication process, resulting in genetic variations, which in turn causes difference in phenotype, the observable traits of organisms. Individuals with traits that are more suitable to their particular environment are more likely to survive and pass their genes to their offspring, comparing to those that are poorly adapted to the environment. This process is termed natural selection. A classic example of evolution is the adaptive radiation of finches on the Galapagos islands. The beak shapes of the finches on different islands diversified to specialize on obtaining the dominant type of food on different islands. Over generations, species evolve as a consequence of the change in the prevalence of advantageous genes. Finally, a new species is formed during the process of speciation. It occurs when enough genetic variations are accumulated between the original population and the descendant population, such that organisms from the two groups are not capable of producing fertile offspring, namely, the two groups have reproductive isolation. \\

It is believed that life emerged from 3.5 billion years ago. The past 540 million years ago (the Phanerozoic Eon) sees rapid development and diversification of life. Throughout evolutionary history, there are records of drastic changes in biodiversity occurring in a relatively short period of time. Radiations are rapid increase in biodiversity. An example is the Cambrian explosion, during which most phylums of multi-cellular organisms appeared. On the contrary, biodiversity experiences rapid decline during mass extinction periods. There is evidence of five mass extinction periods, one of which occurred in Cretaceous and Tertiary, when the dinasours were exterminated and mammals thrived due to their selective advantage given by their fur in the cooling environments. These periods of radiation or extinction are caused by special events, plate motions, as well as change in climate, such as temperature and oxygen level. \\

\subsection{Tectonic shift as a driver of biodiversity change} \label{subsec: two}
The earth crust is composed of a number of tectonic plates undergoing constant changes. The hear generated from the interior of the planet causes the movement of the plates, as a result, earth appears very differently throughout millions of years of history. The plate motion results in new landscapes and new continents, affecting the inhabitants of lives. For instance, the current continents are thought to belong to one super continent called Pangaea, and gradually drifted away from each other since 175 million years ago. \\

As the continents separated, populations from the same species are geographically isolated into their particular environments that is created from the plate motion. Since organisms are under constant selective pressure from the environment, new species are formed in the newly created environments. An example is the distribution of ratite as a result of tectonic movement. 

\section{Source of data} \label{sec:three}
\emph{The fossil}\\
The biodiversity of ancient time periods is represented by fossil records. For this purpose, we use paleodb, which contains 1.3 million occurrences of fossil records. Each record has its unique ID, identified taxonomy, estimated maximal and minimal time (in million years) to today, and their coordinates. \\

\noindent \emph{The tree}\\
All identified species are classified in taxonomy, following the hierarchy of domain, kingdom, phylum, class, order, family, genus and species. Therefore, it can be naturally organized into a tree structure. Conveniently, for each fossil record in PBDB dataset, the names on each rank of the taxonomy is available. The highest taxonomy rank in the PBDB dataset is kingdom, namely the kingdom of Animalia (animal) and Plantae (plants). Since the two kingdoms are under the domain of Eukaryota, a rooted tree of life can be generated from the taxonomy information of all fossil record.\\

\noindent \emph{The map}\\
The geological information is provided by Matthew and colleagues, who traced the coastlines of the continents from 410 million years ago until the current world. 

\section{Aim of the thesis}
This thesis aims at building a web application that visualizes the change in biodiversity through the course of earth history. In the final product, users can view a customizable tree of life during a selected time period under a selected taxonomy. In the meantime, the fossil records belonging to the given tree of life is shown on the map, whose coastlines correspond to that of the selected epoch in ancient time. 

\section{Thesis structure}
Chapter one introduces the topic of data visualization and domain specific background information, and it claims the goal of the thesis. Chapter two analyzes some related work, indicating what could be learned and improved for the visualization to be built. Chapter three outlines the problems to be solved at each stage for building the web application. Chapter four discuses the technical solutions for data processing and web development stage, respectively. In chapter five, the implementation methods for data processing, backend and frontend are listed. Chapter six reports the result of the product. Chapter seven concludes the project and suggests future work.


%=====================================================================
\chapter{Related work} \label{chp:related_work}
%=====================================================================

There already exists an abundance of visualizations on the Internet regarding to tree of life as well as plate motion. This section selects several related work and discusses their inspirations for designing our own visualization on this topic. 

\section{Tree-based visualizations on biodiversity}
OneZoom (\url{https://www.onezoom.org/}, Fig.\ref{fig:OneZoom}) visualizes all known living things in a fractal tree, and it is targeted for both professionals and the masses. Users can zoom into or out of different levels of the tree by scrolling, and can search for a particular location in the tree by its name. 

\begin{figure}[h]
	\centering
	\subfigure[The overview of OneZoom's tree from root. ]{\includegraphics[width=0.45\textwidth]{figures/related_work/onezoom_overview}\label{fig:OneZoom_a}} 
	\hfill
	\subfigure[The view on OneZoom's tree when zoomed onto mammals. ]{\includegraphics[width=0.45\textwidth]{figures/related_work/onezoom_mammal}\label{fig:OneZoom_b}} 
	\caption{Tree of life implementation in OneZoom. }
	\label{fig:OneZoom}
\end{figure}

The fractal design gives the tree a nice appearance, with leaves and nodes resembling the leaves and branches of an actual tree. General information such as pictures or earliest time of a node is also well packed in the interactive image. There is a reset button to return to the root, and a drop down list that provides links to the nodes along the path to root from the current position. These functions are convenient for navigation and similar functionality should be implemented in this thesis. It is claimed that all categorized organisms, which has a total amount of over two million, are included in the tree. Each species in the tree structure are incorporated with a phylogenetic classification, indicating that for a particular species represented by a leaf to reach the root, it visits all of the nodes representing each of its ancestor throughout the evolutionary history. Although how a certain organism evolved from the common ancestor of all life can be rigorously inspected, from the user experience's point of view, the tree is extremely deep. Apart from manually searching for a name, navigating to a desired location in this deep tree is mainly achieved by scrolling and dragging the item of interest to the center of the screen. This makes it difficult for viewers to relocate to a different section of the tree, or to gain an overview of where the current node or leaf sits in the entire tree of life. To reach the root (\ref{fig:OneZoom_a}) from mammal (\ref{fig:OneZoom_b}) while viewing its path, users need to scroll through all upper levels of the similar-looking spiral structures of the tree, which appears to be confusing. A similar tree of life is Lifemap (\url{http://lifemap.univ-lyon1.fr/explore.html}), which ameliorates the navigation difficulty by highlighting the ancestry path from root to an entry (Fig\ref{fig:lifemap}). However, due to the sheer depth of the tree, users still need to travel deep into the tree by scrolling. Rather than demonstrating the exact ancestral paths, the purpose of the tree for this project is to organize each species such that biodiversity can be viewed in an intuitive way. As a consequence, all entries can be organized in only the main taxonomic ranks, limiting the tree's depth to be eight and thus simplifying the navigation process. In addition, to avoid scrolling and dragging, the tree can take a fixed position, while where a node or leaf locates can be shown in a clickable breadcrumb structure that provides an overview of the whole tree. \\

\begin{figure}[h]
	\centering
	\includegraphics[width=0.5\textwidth]{figures/related_work/lifemap}
	\caption{Lifemap enables a "path from root" functionality that highlights the ancestry path from \textit{Homo sapiens} to the root. }
	\label{fig:lifemap}
\end{figure}

A much simplified version of tree of life is Evogeneao (\url{https://www.evogeneao.com/en/explore/tree-of-life-explorer}), which is intended for school education. As shown in Fig\ref{fig:evogeneao}, all species are grouped into clearly colored groups, which gives the viewers a direct and memorable visual impact. A similar design could be adapted in this project. With the tree of life arranged in a half circle that is labeled with time and a few iconic events, users could get a view of how life radiated from the common ancestor into great diversity, and how this process is affected by special events on the planet. Since only a small subset of all organisms are shown, the tree is small and can be entirely shown, thus always giving the audience an overview. However, this web application provides little possibility of exploration due to its limited options for interaction, as the main functionality is to show the common ancestor of two selected items (Fig\ref{fig:evogeneao_b}).

\begin{figure}[h]
	\centering
	\subfigure[The overview of Evogeneao's tree. ]{\includegraphics[width=0.45\textwidth]{figures/related_work/evogeneao_overview}\label{fig:evogeneao_a}} 
	\hfill
	\subfigure[The ancestral relationship between human and songbird. ]{\includegraphics[width=0.45\textwidth]{figures/related_work/evogeneao_relationship}\label{fig:evogeneao_b}} 
	\caption{Tree of life implementation in Evogeneao. }
	\label{fig:evogeneao}
\end{figure}

Finally, there are a wide variety of tree of life visualized for computational biologists. Such visualizations provide rich functionalities for bioinformatic analysis that are well beyond the need of general public and often take efforts to use. An example is iTOL (\url{https://itol.embl.de/itol.cgi},Fig\ref{fig:iTOL}). The design of arranging all leaves on the edge of a circle and all nodes in its interior utilizes space effectively, therefore can be adapted in our implementation. 

\begin{figure}[h]
	\centering
	\includegraphics[width=0.5\textwidth]{figures/related_work/iTOL}
	\caption{Tree of life in iTOL. }
	\label{fig:iTOL}
\end{figure}

\section{Visualizations on tectonic shift}
\begin{figure}[h]
	\centering
	\subfigure[Tectonic shift visualized in Ancient Earth. ]{\includegraphics[width=0.45\textwidth]{figures/related_work/ancient_earth}\label{fig:ancient_earth}} 
	\hfill
	\subfigure[Tectonic shift visualized in Earth Viewer. ]{\includegraphics[width=0.45\textwidth]{figures/related_work/earth_viewer}\label{fig:earth_viewer}} 
	\caption{Example visualizations on tectonic shift. }
	\label{fig:tectonic_shift}
\end{figure}
Current project involves visualizing plate motions in various time period, since biodiversity information is shown on the world map during the select time period of interest. Two typical tectonic shift visualizations are Ancient Earth (\url{https://dinosaurpictures.org/ancient-earth#240}, Fig\ref{fig:ancient_earth}) and Earth Viewer (\url{https://www.biointeractive.org/classroom-resources/earthviewer}, Fig\ref{fig:earth_viewer}). Both of them show the coastlines of the continents at a select time point on a three-dimensional globe. Ancient Earth allows users to pick a particular time point from a list of options and search for the ancient location of a city. Earth Viewer provides more detailed information on geological time, and selectable time points are incremented at every five million years on the scroll bar. Time periods are organized in a hierarchical structure with names, which can be adapted in this project for easy navigation. When clicking on a time period, instead of directing the user to the selection, a pop-up window containing a brief introduction of the time period appears, which is counter-intuitive. An improvement is to show the earth at the middle of the selected time period upon clicking.

%=====================================================================
\chapter{Problem statement} \label{chp:problem_statement}
%=====================================================================
The data visualization project is divided into several sub-problems as follows.

\begin{itemize}
	\item \textbf{Data processing}
	\begin{itemize}
		\item Process the PBDB dataset to extract fossil records 
		\item Extract a tree of life from the taxonomy information of the fossil record 
		\item Trace the coordinates of each fossil record to ancient time
	\end{itemize} 
	\item \textbf{Database setup}: store the fossil data, coordinate data and the taxonomy data into a database
	\item \textbf{Back-end building}: enable data queries from the database upon users' request from the front-end 
	\item \textbf{Front-end building}: visualize interactive components for users to explore biodiversity through deep time
	\begin{itemize}
		\item A time component that lists ancient epochs that users can select
		\item A tree component that shows the tree of life constructed by the fossil records found in the selected time period, under the selected taxonomy. 
		\item A map component that shows the shape of the continents in the selected time period, and the locations of the fossil records that constitutes the tree of life in the above component.    
	\end{itemize} 
\end{itemize} 

%=====================================================================
\chapter{Technical solution} \label{chp:technical_solution}
%=====================================================================
\section{Overview}\label{sec:tec_overview}
This section explains the technical solutions to the problems stated in chapter \ref{chp:problem_statement}. To summarize, the project is divided into data processing in section \ref{sec:tec_data_processing}, database setup in section \ref{sec:tec_database_setup}, back-end development in section \ref{sec:tec_backend} and front-end development in section \ref{sec:tec_frontend}, and the workflow of the project follows the same order. 

\begin{figure}[h]
	\centering
	\includegraphics[width=0.9\textwidth]{figures/technical_solution/architecture}
	\caption{An overview of the architecture solution for the project. The user interface consists of a map, a tree and a time table. The global states store tree and time parameters that are customized by users, and the current tree nodes attached with fossil ID sent from database. The back-end resolves queries on tree, fossil locations and map, before sending data back to front-end, where the data is visualized. }
	\label{fig:architecture}
\end{figure}

Diagram \ref{fig:architecture} proposes a structure of the designed visualization system. The front-end of the web application has a user interface and global states section that stores variables accessible to the components of the user interface. The back-end stores data either in database or as static files, and it responds to the data requests by the front-end, by sending data back to global states section for storage, or directly to the user interface for visualization. Map, tree and time control are the three components that users can view and interact with. The time control panel allows users to select a time period to either view the tree of life, or track the fossil location change due to plate motion. The \emph{time parameters} from global states section listen to user-induced changes in the time control panel, and triggers request of map data, tree data (together with tree parameters stored in global states) or fossil location data (together with fossil ID information stored in global states). The map component receives 1, coastline boundary data in GeoJSON format and 2, fossil location data in database at back-end, to render world map and fossils in a selected time period. The tree component shows the tree of life stored in global states. When users customize the tree view, the tree parameters are updated, and new data on tree nodes with attached fossil IDs is queried from the database. The data is returned to the tree and fossil ID storage in global states, which in turn sends only the tree information to the tree component to be rendered as the new tree. 

\section{Data processing} \label{sec:tec_data_processing}
\subsection{Raw data parsing} \label{subsec:raw_data_parsing}
The data source for fossil records and the tree of life is provided by Paleobiology Database (PBDB). The raw data consists of two csv files that contain 1240202 animal fossil records and 105918 plant fossil records, respectively. Since the two files share the same data attributes, the data is concatenated into one file. Only the attributes that are relevant to this project are parsed, which are listed below.\\

\noindent Extracted attributes in the PBDB dataset:
\begin{itemize}
	\item \textit{occurrence\_no}: a unique ID code for a fossil record
	\item \textit{accepted\_name}: the accepted name of a fossil record
	\item \textit{accepted\_rank}: the taxonomic rank in which the accepted\_name belongs to 
	\item \textit{phylum}: the phylum name of the fossil record
	\item \textit{class}: the class name of the fossil record
	\item \textit{order}: the order name of the fossil record
	\item \textit{family}: the family name of the fossil record
	\item \textit{genus}: the genus name of the fossil record
	\item \textit{max\_ma}: the oldest estimated time of the fossil record, in million years
	\item \textit{min\_ma}: the earliest estimated time of the fossil record, in million years
	\item \textit{lng}: the longitude of the fossil record's found location 
	\item \textit{lat}: the latitude of the fossil record's found location 
	
\end{itemize}

In addition, we append a new attribute to the dataset to keep note of which one of the two files a fossil record is originated from: 

\begin{itemize}
	\item \textit{kingdom}: the kingdom name of the fossil record	
\end{itemize}

The kingdom of Animalia and Plantae is assigned to each fossil record, depending on whether it is from the animal or plant file, respectively. \\

By observing the available attributes in the raw data, we can determine how it can be used for an interactive visualization. The biodiversity of a given time period is represented by 1, a dynamic tree of life formed by the fossil records during that time period, and 2, the locations of the fossils on a world map of that time period. The \emph{occurrence\_no} uniquely identifies a fossil record. The \emph{max\_ma} and \emph{min\_ma} together describe during which time period(s) the organism was alive. The \emph{lat} and \emph{lng} are required to calculate the location of a record during an ancient time period. Most importantly, a tree of life is needed to host each fossil record on a specific node. To generate a tree structure, information on parent-child relationship is essential. This relationship is represented in the names on each of the taxonomic ranks, in the descending order of \emph{domain}, \emph{kingdom}, \emph{phylum}, \emph{class}, \emph{order}, \emph{family}, \emph{genus}, \emph{species}. The raw data lacks information on \emph{domain} and \emph{species}. However, since all records are either animals or plants, they all belong to the common domain of Eukaryota. When the value of \emph{accepted\_rank} is species, the species name is found in \emph{accepted\_name}. Therefore, the raw data has sufficient information to construct a tree of life with the eight above-mentioned taxonomic ranks as its levels.

The datasets on fossils and tree are created from this parsed raw data. In the next steps, we first prepare the dataset that contains all fossil points, since the tree dataset is created based on it. 

\subsection{Fossil point dataset preparation}
\label{subsec: fossil_preparation}
There are two datasets on fossil records. This section produces \emph{fossil point}, the dataset on the taxonomy and the range of age for each record, and sub-section \ref{subsec:coordinate_conversion} produces \emph{fossil location}, the dataset that stores the coordinates of each fossil record on various time periods in history. \\

The purpose of \emph{fossil point} dataset is to accurately describe a given fossil record in relation to the tree of life, without considering its location at a given time period. The parsed raw data already provides \emph{occurrence\_no}, a unique ID code, as well as \emph{max\_ma} and \emph{max\_ma}, which indicate its range of age in million years. However, one more attribute is required to describe its taxonomy, such that it can be attached to a particular node in the tree of life to be constructed in the next step. This attribute is essentially the unique identifier of a node in the tree of life, and will also serve as the link between the fossil point dataset and the tree dataset. Although it is natural to consider \emph{accepted\_name} and \emph{accepted\_rank} in the parsed raw data for this purpose, they cannot be used for the following reasons. Firstly, \emph{accepted\_name} and \emph{accepted\_rank} does not correlate well with the names on the taxonomic ranks, as some \emph{accepted\_rank} does not belong to any of the eight ranks that will be used in the tree of life. As a consequence, some fossil records cannot be assigned to the tree, since the \emph{accepted\_name} values corresponding to the \emph{accepted\_rank} will not be found in the tree. Secondly, the \emph{accepted\_rank} and \emph{accepted\_name} attributes are not unique. For instance, the genus "Banksia" is both an animal taxonomy and a plant taxonomy. In such cases, fossil records can be associated to a incorrect branch if only these two pieces of information is used. Thirdly, many names in the taxonomic ranks are missing. This leads to ambiguity when attaching a particular fossil to the tree of life, when there exists a rank above the \emph{accepted\_rank} with missing names. An example is shown below.

\begin{table}[h]
	\centering
	\begin{tabular}{|l|l|l|l|l|l|l|l|l|l|}
		\hline
		occurrence\_no &  accepted\_name & accepted\_rank & kingdom &	phylum & class & order & family & genus & ... \\ \hline
		3285 &  Tetradium & genus & Plantae & Rhodophyta & NaN & NaN & NaN & Tetradium & ... \\ \hline
	\end{tabular}
	\caption{An example row of parsed raw data with missing information.}
	\label{tab:missing_information}
\end{table}

With only the \emph{accepted\_name} "Tetradium" on the \emph{accepted\_rank} "genus", we are uncertain about which class, order or family to associate this fossil record to under the phylum of "Rhodophyta". Extra knowledge in taxonomy is needed to retrieve the missing names on the path from "Rhodophyta" to "Tetradium" and this challenge is beyond the scope of the current project. The current solution only attaches the fossil record "3285" to the phylum of "Rhodophyta" and ignores its taxonomy below the phylum level. \\

To solve these three issues and provide a method for uniquely identifying a fossil record to the tree without ambiguity, the attributes \emph{path\_from\_root} and \emph{rank} are introduced. 

\begin{itemize}
	\item \textit{path\_from\_root}: a comma separated string indicating the path from the root Eukaryota to the name on lowest possible taxonomic rank with no missing names in between
	\item \textit{rank}: the lowest taxonomic rank of \textit{path\_from\_root}
\end{itemize}

Since a path begins from the common root Eukaryota and terminates when reaching the lowest rank level "species", or encountering a missing name, the example in Table \ref{tab:missing_information} has "Eukaryota, Plantae, Rhodophyta" as \emph{path\_from\_root} and "phylum" as \emph{rank}. All names in \emph{path\_from\_root} are guaranteed to be present in the tree, since the names are taken from the same taxonomy information that is used to construct the tree. The path from the common root Eukaryota describes the exact location of a node in the tree, which serves as the node's unique identifier. In addition, the fossils associated to the subtree under a certain taxonomy can be easily obtained by string matching this path using regular expression. \\

To calculate \emph{path\_from\_root} and \emph{rank}, the following attributes are used:  \emph{domain}, \emph{kingdom}, \emph{phylum}, \emph{class}, \emph{order}, \emph{family}, \emph{genus}, \emph{accepted\_name}, \emph{accepted\_rank}. Instead of iterating over each row in the parsed raw data, we first aggregate the rows when those values are identical and then perform the iteration, before joining the result with the original data (similar to an "inner join" operation in SQL), in order to avoid repetitive computation. \\

At each iteration process on the aggregated data, we initialize the \emph{path\_from\_root} as "Eukaryota" and visit each taxonomic rank from \emph{kingdom} to \emph{genus}, while appending a comma and the name on each rank to the end of the string and updating the current rank. The iteration terminates at the first missing name, when both \emph{path\_from\_root} and \emph{rank} are those of the missing name's parent. After visiting \emph{genus}, the iteration finishes when the \emph{accepted\_rank} is "genus". Otherwise, the \emph{accepted\_rank} is "species" or "subspecies". Under this circumstance, we append the value of \emph{accepted\_name} to \emph{path\_from\_root}, update \emph{rank} to be the value of \emph{accepted\_rank}, and terminate the iteration. The name in the path after genus can thus be either a species or a subspecies, which is acceptable, since a subspecies name already includes the species name (\emph{i.e.}, the species name of the subspecies "Argopecten circularis impostor" is "Argopecten circularis"). \\

It is important to note that the columns \emph{path\_from\_root} and \emph{rank} are precursors to the tree dataset, since they represent all nodes with at least one fossil record attached. The unique values of the two columns are used for further processing in sub-section \ref{subsec:tree_preparation}. In the meantime, the aggregated data with \emph{path\_from\_root} appended are joined with parsed raw data, such that every fossil record has a path. Finally, the \emph{fossil point} dataset is generated, by extracting the following attributes: 

\begin{itemize}
	\item \textit{occurrence\_no}
	\item \textit{path\_from\_root}
	\item \textit{max\_ma}
	\item \textit{min\_ma}
\end{itemize}


\subsection{Tree dataset preparation} \label{subsec:tree_preparation}
The purpose of the tree dataset is to describe every node on the tree of life that can also be easily searched. The following attributes are computed: \\

\begin{itemize}
	\item \textit{path\_from\_root}: a comma separated string indicating the path from the root Eukaryota to the name on lowest possible taxonomic rank with no missing names in between
	\item \textit{parent}: the \textit{path\_from\_root} of its parent
	\item \textit{name}: the name of the node
	\item \textit{is\_leaf}: a boolean value that indicates if the node is a leaf node
	\item \textit{max\_ma}: the earliest age in million years of the fossils in the subtree of this node
	\item \textit{min\_ma}: the latest age in million years of the fossils in the subtree of this node
\end{itemize}

For each node, we need a unique identifier, which is provided by \emph{path\_from\_root} as discussed in the above sub-section \ref{subsec: fossil_preparation}. The last \emph{name} of the path is also extracted as an attribute, enabling a more convenient node search when users type a name in the search bar. The node's parent attribute allows searching for all children of a given node. The \emph{is\_leaf} attribute indicates directly whether or not a node has children without searching for them. Lastly, \emph{max\_ma} and \emph{min\_ma} provide a quick way to tell if a node is present during a given time period without searching for fossils attached to its subtree. \\

The tree data contains all nodes and leaves of the tree of life formed by the taxonomy of the fossil records. We initialize the tree data frame by extracting all the unique \emph{path\_from\_root} values from the processed fossil data from section \ref{subsec:fossil_data_processing}. However, there are more nodes to include in the tree. For each entry in the tree represented by its \emph{path\_from\_root}, we must also include each of the upstream ancestor in the tree. For instance, when the item "a,b,c,d" is included, its ancestors "a,b,c", "a,b" and "a" are also present in the tree. Therefore, the next step is to iterate through each of the unique \emph{path\_from\_root} and append the additional nodes. \\

The next steps are to append the \emph{unique\_name}, \emph{parent} and \emph{children} values for each \emph{path\_from\_root}. Similar to section \ref{subsec:fossil_data_processing}, we compute \emph{unique\_name} from \emph{path\_from\_root}, by firstly extracting \emph{name}, the name on the lowest rank, and assign either \emph{name} to \emph{unique\_name} if \emph{name} is unique, or assign \emph{path\_from\_root} to \emph{unique\_name} if otherwise. \\

The attribute \emph{parent} can be obtained from slicing the \emph{path\_from\_root} from start to its last comma (the \emph{path\_from\_root} value of its parent), and assigning the corresponding \emph{unique\_name} value to \emph{parent}. Lastly, we iterate through each \emph{unique\_name} and append the list of \emph{unique\_name} values of all the items whose \emph{parent} equals the current \emph{unique\_name} in each iteration.

\subsection{Conversion of modern coordinates to ancient coordinates} \label{subsec:coordinate_conversion}
To convert coastlines and fossil locations to ancient time, we use the model proposed by Matthew. \\

Geological time period data is hierarchical. The hierarchy contains five layers, with the top most layer Phanerozoic. A time period in a deeper layer is shorter. The ancient year that the coordinates are translated into is the middle value rounded to the integer, between the start and end years of the smallest time period, i.e., layer five. 

\section{Database setup} \label{sec:tec_database_setup}
The data is stored in a MongoDB database. As shown in above, the fossil data are split into two collections, FossilPoint and FossilLocation. FossilPoint stores the ID, uniqueName,pathFromRoot, maxma and minma of the fossil records. FossilLocation stores the ID, a time point and its coordinate at that time point. The tree data is stored in one collection, TreeNode, containing the uniqueName, pathFromRoot, parent and hasChildren.
 
\section{Back-end development} \label{sec:tec_backend}
The back-end is mainly responsible for processing the data queries requested by user actions from the front-end. There are two main queries to carry out: get a tree with fossil IDs attached to each node, and get the fossil locations of the fossil IDs at a given time in history. 

\section{Front-end development} \label{sec:tec_frontend}
The map is rendered from static files.\\

The fossil locations are rendered from the results of the fossil location query. \\

The tree is rendered from the tree that is retrieved from the back-end and stored in the global state. 

%=====================================================================
\chapter{Implementation} \label{chp:implementation}
%=====================================================================
\section{Code structure} \label{sec:implementation_code_structure}
\section{Data process stacks}\label{sec:implementation_data_processing}
\section{Web development stacks}\label{sec:implementation_web_development}

%=====================================================================
\chapter{Results} \label{chp:results}
%=====================================================================
\section{Data analysis} \label{sec:result_data_analysis}
\subsection{fossil data} \label{subsec:data_analysis_fossil}
\subsection{tree data} \label{subsec:data_analysis_tree}

\section{Data visualization} \label{sec:result_data_visualization}
\subsection{Time selector component} \label{subsec:data_visualization_time}
\subsection{Map component} \label{subsec:data_visualization_map}
\subsection{Tree component} \label{subsec:data_visualization_component}
%=====================================================================
\chapter{Conclusion} \label{chp:discussion_conclusion}
%=====================================================================



\section{Limitations and Future Work}

%=====================================================================
\chapter*{Acknowledgements}
%=====================================================================


\bibliographystyle{alpha}
\bibliography{references}

\listoffigures
\listoftables
%=====================================================================
\appendix
\chapter{Appendix}
The appendix of this thesis is stored on a DVD which is attached to the end of the printed document. It includes the follow items: 
\begin{enumerate}
    \item \textbf{code/} The full application code. 
    \item \textbf{report/} The complete report as latex source
    \item \textbf{Masterarbeit.pdf} This report in its digital version
    \item{Abstract.txt} The abstract to this thesis in English
    \item {Zusfsg.txt} The abstract to this thesis in German
\end{enumerate}

%=====================================================================
\end{document}

